presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayCumSum.java

```diff
@@ -0,0 +1,86 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.block.RunLengthEncodedBlock;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.OperatorDependency;
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.function.TypeParameter;
+
+import java.lang.invoke.MethodHandle;
+
+import static com.facebook.presto.common.function.OperatorType.ADD;
+import static com.facebook.presto.common.type.TypeUtils.readNativeValue;
+import static com.facebook.presto.common.type.TypeUtils.writeNativeValue;
+import static com.facebook.presto.util.Failures.internalError;
+
+@Description("Get the cumulative sum array for the input array")
+@ScalarFunction(value = "array_cum_sum", deterministic = true)
+public final class ArrayCumSum
+{
+    private ArrayCumSum() {}
+
+    @TypeParameter("T")
+    @SqlType("array(T)")
+    public static Block sum(
+            @OperatorDependency(operator = ADD, argumentTypes = {"T", "T"}) MethodHandle addFunction,
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block arrayBlock)
+    {
+        int positionCount = arrayBlock.getPositionCount();
+        BlockBuilder resultBuilder = elementType.createBlockBuilder(null, positionCount);
+        if (positionCount == 0) {
+            return resultBuilder.build();
+        }
+
+        if (arrayBlock.isNull(0)) {
+            return RunLengthEncodedBlock.create(elementType, null, positionCount);
+        }
+
+        elementType.appendTo(arrayBlock, 0, resultBuilder);
+        if (arrayBlock.mayHaveNull()) {
+            int pos = 1;
+            for (; pos < positionCount; ++pos) {
+                if (arrayBlock.isNull(pos)) {
+                    break;
+                }
+                try {
+                    writeNativeValue(elementType, resultBuilder, addFunction.invoke(readNativeValue(elementType, resultBuilder, pos - 1), readNativeValue(elementType, arrayBlock, pos)));
+                }
+                catch (Throwable throwable) {
+                    throw internalError(throwable);
+                }
+            }
+            for (; pos < positionCount; ++pos) {
+                resultBuilder.appendNull();
+            }
+        }
+        else {
+            for (int pos = 1; pos < positionCount; ++pos) {
+                try {
```

Comment:
We can extract `try {write;} catch { error }` in a function as we use it 2 times